src/poll.rs

```diff
@@ -68,95 +113,155 @@ pub struct Registration {
+    // Node state, see struct docs for `ReadinessState`
     //
-    // Next node in ownership tracking queue
-    next_all_nodes: Option<Box<ReadinessNode>>,
+    // This variable is the primary point of coordination between all the
+    // various threads concurrently accessing the node.
+    state: AtomicState,
 
-    // Previous node in the owned list
-    prev_all_nodes: ReadyRef,
-
-    // Data set in register / reregister functions and read in `Poll`. This
-    // field should only be accessed from the thread that owns the `Poll`
-    // instance.
-    registration_data: UnsafeCell<RegistrationData>,
-
-    // ===== Fields accessed by any thread ====
+    // The registration token cannot fit into the `state` variable, so it is
+    // broken out here. In order to atomically update both the state and token
+    // we have to jump through a few hoops.
+    //
+    // First, `state` includes `token_read_pos` and `token_write_pos`. These can
+    // either be 0, 1, or 2 which represent a token slot. `token_write_pos` is
+    // the token slot that contains the most up to date registration token.
+    // `token_read_pos` is the token slot that `poll` is currently reading from.
+    //
+    // When a call to `update` includes a different token than the one currently
+    // associated with the registration (token_write_pos), first an unused token
+    // slot is found. The unused slot is the one not represented by
+    // `token_read_pos` OR `token_write_pos`. The new token is written to this
+    // slot, then `state` is updated with the new `token_write_pos` value. This
+    // requires that there is only a *single* concurrent call to `update`.
     //
+    // When `poll` reads a node state, it checks that `token_read_pos` matches
+    // `token_write_pos`. If they do not match, then it atomically updates
+    // `state` such that `token_read_pos` is set to `token_write_pos`. It will
+    // then read the token at the newly updated `token_read_pos`.
+    token_0: UnsafeCell<Token>,
+    token_1: UnsafeCell<Token>,
+    token_2: UnsafeCell<Token>,
+
     // Used when the node is queued in the readiness linked list. Accessing
     // this field requires winning the "queue" lock
-    next_readiness: ReadyRef,
+    next_readiness: AtomicPtr<ReadinessNode>,
 
-    // The set of events to include in the notification on next poll
-    events: AtomicUsize,
+    // Ensures that there is only one concurrent call to `update`.
+    //
+    // Each call to `update` will attempt to swap `update_lock` from `false` to
+    // `true`. If the CAS succeeds, the thread has obtained the update lock. If
+    // the CAS fails, then the `update` call returns immediately and the update
+    // is discarded.
+    update_lock: AtomicBool,
 
-    // Tracks if the node is queued for readiness using the MSB, the
-    // rest of the usize is the readiness delay.
-    queued: AtomicUsize,
+    // Number of outstanding Registration nodes
+    num_registration: AtomicUsize,
```

Comment:
Reading over all this again, I think this field may no longer be used? It's never incremented and is only decremented in `Drop for Registration`, so I wonder if this was added at a point where it was intended that `Registration` is `Clone`?